src/hasSubObject.ts

```diff
@@ -1,7 +1,15 @@
-import type { Simplify } from "type-fest";
+import type { Merge } from "type-fest";
+import type { Branded } from "./internal/types";
 import { isDeepEqual } from "./isDeepEqual";
 import { purry } from "./purry";
 
+declare const HAS_SUB_OBJECT_BRAND: unique symbol;
+
+type HasSubObjectGuard<T, S> = Branded<
```

Comment:
I'm still not sure about the name, I prefer a more descriptive name and not just a formulaic one. Maybe "RefinedObject", e.g. `hasSubObject<T, S>(data: T, subObj: S): data is RefinedObject<T, S>`, I feel this is the most readable version of the name...

@cjquines , wdyt?

Reply:
@eranhirsch I left some options here - https://github.com/remeda/remeda/pull/683#discussion_r1629100613. What do you think about `WithSubObject`? `RefinedObject` sounds good too.

Reply:
maybe something like

```ts
type WithSubObject<T> = Branded<T, typeof HAS_SUB_OBJECT_BRAND>;

function hasSubObject<T, S>(data: T, subObj: S): data is WithSubObject<Merge<T, S>>;
```

that way we still get nice hover info

<img width="739" alt="image" src="https://github.com/remeda/remeda/assets/3197784/704597c6-1f09-4821-8555-2e9968e388b8">

Reply:
not sure how that reads though... `obj` doesn't just have  `{ b?: number; readonly a: "a"; }` with a subobject it _is_ that

in this case we want something small and ignorable, like `Is` or `Wrap`

(the ideal would be `data is Simplify<WithSubObject<T, S>>`, but that doesn't typecheck?)

Reply:
@cjquines 

> not sure how that reads though... obj doesn't just have { b?: number; readonly a: "a"; } as a subobject type. it is that type

Yeah, in this case it's probably better to have `HasSubObjectGuard`, `HasSubObjectBranded`, `RefinedObject`, `BrandedObject`.

> in this case maybe we want something small and ignorable, like Is or Wrap? read: obj has the type Is<{ b?: number; readonly a: "a"; }>?

We can use `Branded` but I'm not sure about this since we also have an internal type with such name.

> the ideal would be data is Simplify<WithSubObject<T, S>>, but that doesn't typecheck?

We can make it typecheck (even with `Merge`). But in this case, I would probably remove `Merge` at all and revert `Simplify` usage and use it as the following:

![image](https://github.com/remeda/remeda/assets/10157660/e309f5f4-aade-4ef7-9620-56799ace1542)

![image](https://github.com/remeda/remeda/assets/10157660/1ef784a4-5146-46d8-8d74-da33bfbe0c4c)


In another PR, I had to use `Simplify` instead of `Merge` anyway - https://github.com/remeda/remeda/pull/683#discussion_r1620560556.

Reply:
let's stick with that then—the simplify option looks better to me?